ambry-tools/src/main/java/com.github.ambry/clustermap/HelixBootstrapUpgradeTool.java

```diff
@@ -0,0 +1,466 @@
+        DataNodeId node = replicaId.getDataNodeId();
+        String instanceName = node.getHostname() + "_" + node.getPort();
+        InstanceConfig instanceConfig = dcAdmin.getInstanceConfig(clusterName, instanceName);
+        List<String> currentSealedPartitions = instanceConfig.getRecord().getListField("SEALED");
+        if (isSealed && !currentSealedPartitions.contains(partitionName)) {
+          List<String> newSealedPartitionsList = new ArrayList<>(currentSealedPartitions);
+          newSealedPartitionsList.add(partitionName);
+          instanceConfig.getRecord().setListField("SEALED", newSealedPartitionsList);
+          dcAdmin.setInstanceConfig(clusterName, instanceName, instanceConfig);
+        } else if (!isSealed && currentSealedPartitions.contains(partitionName)) {
+          List<String> newSealedPartitionsList = new ArrayList<>(currentSealedPartitions);
+          newSealedPartitionsList.remove(partitionName);
+          instanceConfig.getRecord().setListField("SEALED", newSealedPartitionsList);
+          dcAdmin.setInstanceConfig(clusterName, instanceName, instanceConfig);
+        }
+      }
+    }
+  }
+
+  /**
+   * Adds all partitions to every datacenter with replicas in nodes as specified in the static clustermap (unless it
+   * was already added).
+   *
+   * The assumption is that in the static layout, every partition is contained in every colo. We make this assumption
+   * to ensure that partitions are grouped under the same resource in all colos (since the resource id is not
+   * something that is present today in the static cluster map). This is not a strict requirement though, but helps
+   * ease the logic.
+   *
+   * Note: We ensure that the partition names are unique in the Ambry cluster even across resources.
+   */
+  private void addNewAmbryPartitions(List<Partition> partitions, String resourceName) {
+    // In the future, a resource may be used to group together partitions of a container. For now, multiple
+    // resources are created and partitions are grouped under these resources upto a maximum threshold.
+    if (partitions.isEmpty()) {
+      throw new IllegalArgumentException("Cannot add resource with zero partitions");
+    }
+    for (Map.Entry<String, HelixAdmin> entry : adminForDc.entrySet()) {
+      String dcName = entry.getKey();
+      HelixAdmin dcAdmin = entry.getValue();
+      AutoModeISBuilder resourceISBuilder = new AutoModeISBuilder(resourceName);
+      int numReplicas = 0;
+      resourceISBuilder.setStateModel("LeaderStandby");
+      for (Partition partition : partitions) {
+        String partitionName = Long.toString(partition.getId());
+        boolean sealed = partition.getPartitionState().equals(PartitionState.READ_ONLY);
+        List<ReplicaId> replicaList = getReplicasInDc(partition, dcName);
+        numReplicas = replicaList.size();
+        String[] instances = updateInstancesAndGetInstanceNames(dcAdmin, partitionName, replicaList, sealed);
```

Comment:
why? `partitionName` and `numReplicas` are required after the call to `updateInstancesAndGetInstanceNames()`.